### ISSUE TITLE
iteration vs epochs

### ISSUE THREAD

**Issue by vince62s** [facebookresearch/fairseq-py#108]
Hey guys,

When dealing with a big dataset it's a bit a pain to manage epochs.
Do you intend to allow iterations schedule for learning rate decay ?
cheers, great work.

**Comment by myleott** [comment]
This has been added with the most recent update. Please see the InverseSquareRootScheduler as an example: https://github.com/facebookresearch/fairseq-py/blob/master/fairseq/optim/lr_scheduler/inverse_square_root_schedule.py#L69-L76

**Comment by vince62s** [comment]
oh ok thanks.
wha triggers the end of training then? is there a max_updates parameter ? still max_epochs ? 
thanks

**Comment by myleott** [comment]
Yes it's still based on max_epochs. If you want to stop training after a certain number of updates, you can update singleprocess_train.py [1] and add a condition that breaks once `trainer.get_num_updates()` surpasses your target.

[1] https://github.com/facebookresearch/fairseq-py/blob/master/singleprocess_train.py#L137-L154

**Comment by edunov** [comment]
Another option would be to use --sample-without-replacement NUM_UPDATES that lets you redefine the size of an epoch. This way you can keep learning rate scheduler the same.

**Comment by vince62s** [comment]
great this is what I need.
but beside this, is there any benefit using the invsqrt lr scheduler ? did you see some good from that ?

**Comment by myleott** [comment]
The inverse sqrt schedule was suggested in the "Attention is All You Need" paper, where they use it with the Adam optimizer. The advantage is that it anneals the learning rate gradually during training, but in practice it performs similarly to NAG (the default), as long as you anneal the learning rate near the end of training.
